File: app/src/main/java/com/unsupportedpastels/hermesandroid/connection/PromptSubmissionLifecycle.kt
package com.unsupportedpastels.hermesandroid.connection

import com.unsupportedpastels.mercury.core.transcript.PromptSubmissionDecision
import com.unsupportedpastels.mercury.core.transcript.PromptSubmissionPolicy

/**
 * Native owner for one per-session prompt.submit admission attempt.
 *
 * The event stream and the correlated prompt.submit response are independent
 * asynchronous results. A terminal event may therefore arrive while the ACK
 * is still pending. Keep the attempt identity and terminal-before-ACK fact
 * together so a late callback cannot settle a replacement draft.
 */
internal class PromptSubmissionLifecycle {
    /** Each Attempt instance is an identity token; fields are never used to settle callbacks. */
    internal class Attempt(
        val draft: String,
        val attachmentIds: List<String>,
    )

    internal enum class Phase {
        Idle,
        AwaitingAcceptance,
        TerminalBeforeAcceptance,
    }

    internal sealed interface TerminalEffect {
        data class ReleasedPending(
            val draft: String,
            val attachmentIds: List<String>,
        ) : TerminalEffect

        data object Ignored : TerminalEffect
    }

    internal sealed interface Resolution {
        data class Accepted(
            val draft: String,
            val attachmentIds: List<String>,
            val terminalAlreadyObserved: Boolean,
        ) : Resolution

        data class RestoreDraft(val draft: String) : Resolution

        data object AuthoritativeTerminal : Resolution

        data object Stale : Resolution
    }

    var phase: Phase = Phase.Idle
        private set

    private var currentAttempt: Attempt? = null

    /** Starts a new attempt unless the current attempt still awaits its ACK. */
    fun begin(draft: String, attachmentIds: List<String>): Attempt? {
        if (phase == Phase.AwaitingAcceptance) return null
        val attempt = Attempt(
            draft = draft,
            attachmentIds = attachmentIds.toList(),
        )
        currentAttempt = attempt
        phase = Phase.AwaitingAcceptance
        return attempt
    }

    /** Releases admission only for a terminal event on the current attempt. */
    fun observeTerminal(): TerminalEffect {
        val attempt = currentAttempt ?: return TerminalEffect.Ignored
        if (phase != Phase.AwaitingAcceptance) return TerminalEffect.Ignored
        if (PromptSubmissionPolicy.onTerminalEvent(awaitingAcceptance = true) !=
            PromptSubmissionDecision.RELEASE_PENDING
        ) {
            return TerminalEffect.Ignored
        }
        phase = Phase.TerminalBeforeAcceptance
        return TerminalEffect.ReleasedPending(
            draft = attempt.draft,
            attachmentIds = attempt.attachmentIds,
        )
    }

    /**
     * Settles one asynchronous acknowledgement. An attempt identity mismatch
     * is a strict no-op; it cannot touch the current replacement attempt.
     */
    fun resolve(attempt: Attempt, accepted: Boolean): Resolution {
        val current = currentAttempt
        if (current == null || current !== attempt) return Resolution.Stale

        val terminalAlreadyObserved = phase == Phase.TerminalBeforeAcceptance
        val decision = PromptSubmissionPolicy.onAcknowledgement(
            authoritativeTerminal = terminalAlreadyObserved,
            accepted = accepted,
        )
        currentAttempt = null
        phase = Phase.Idle

        return when (decision) {
            PromptSubmissionDecision.ACCEPTED -> Resolution.Accepted(
                draft = current.draft,
                attachmentIds = current.attachmentIds,
                terminalAlreadyObserved = terminalAlreadyObserved,
            )
            PromptSubmissionDecision.RESTORE_DRAFT -> Resolution.RestoreDraft(current.draft)
            PromptSubmissionDecision.IGNORE,
            PromptSubmissionDecision.RELEASE_PENDING,
            -> Resolution.AuthoritativeTerminal
        }
    }
}

File: ios/Mercury/MercuryKit/Chat/PromptSubmissionLifecycle.swift
import Foundation
import MercuryCore

/// Native owner for one prompt.submit admission attempt.
///
/// The WebSocket event stream and the correlated prompt.submit response are
/// independent asynchronous results. A terminal event can therefore arrive
/// while the RPC acknowledgement is still pending. This helper keeps the
/// attempt identity and terminal-before-ack fact together so a late callback
/// cannot settle a newer draft.
struct PromptSubmissionLifecycle: Sendable, Equatable {
    struct Attempt: Sendable, Equatable {
        fileprivate let id: UInt64
        fileprivate let draft: String
        fileprivate let hostReferenceIDs: [String]
    }

    enum Phase: Equatable, Sendable {
        case idle
        case awaitingAcceptance
        case terminalBeforeAcceptance
    }

    enum TerminalEffect: Equatable, Sendable {
        case releasedPending(hostReferenceIDs: [String])
        case ignored
    }

    enum Resolution: Equatable, Sendable {
        case accepted(hostReferenceIDs: [String])
        case restoreDraft(String)
        case authoritativeTerminal
        case stale
    }

    private(set) var phase: Phase = .idle
    private var nextAttemptID: UInt64 = 0
    private var currentAttempt: Attempt?

    /// Starts a new admission attempt. A second attempt is rejected until the
    /// current one is authoritatively terminal or has been acknowledged.
    mutating func begin(draft: String, hostReferenceIDs: [String] = []) -> Attempt? {
        guard phase != .awaitingAcceptance else { return nil }
        nextAttemptID &+= 1
        let attempt = Attempt(
            id: nextAttemptID,
            draft: draft,
            hostReferenceIDs: hostReferenceIDs
        )
        currentAttempt = attempt
        phase = .awaitingAcceptance
        return attempt
    }

    /// Records a terminal message/error for the currently awaiting attempt.
    /// The terminal event releases the UI gate before prompt.submit's ACK, but
    /// does not discard the attempt: a late ACK still needs an identity check.
    /// Reference ids are returned so the caller can remove only the accepted
    /// attempt's host references before a follow-up send reuses the composer.
    mutating func observeTerminal() -> TerminalEffect {
        guard let currentAttempt, phase == .awaitingAcceptance else {
            return .ignored
        }
        let decision = MercuryCore.PromptSubmissionPolicy.shared.onTerminalEvent(
            awaitingAcceptance: true
        )
        guard decision == .releasePending else { return .ignored }
        phase = .terminalBeforeAcceptance
        return .releasedPending(hostReferenceIDs: currentAttempt.hostReferenceIDs)
    }

    /// Settles one asynchronous acknowledgement. Stale attempts are a strict
    /// no-op: they cannot clear the current gate, surface an old error, restore
    /// old text, or remove references staged for a newer submission.
    mutating func resolve(attempt: Attempt, accepted: Bool) -> Resolution {
        guard let currentAttempt, currentAttempt.id == attempt.id else {
            return .stale
        }

        let decision = MercuryCore.PromptSubmissionPolicy.shared.onAcknowledgement(
            authoritativeTerminal: phase == .terminalBeforeAcceptance,
            accepted: accepted
        )
        self.currentAttempt = nil
        phase = .idle

        switch decision {
        case .accepted:
            return .accepted(hostReferenceIDs: currentAttempt.hostReferenceIDs)
        case .restoreDraft:
            return .restoreDraft(currentAttempt.draft)
        case .ignore:
            return .authoritativeTerminal
        case .releasePending:
            // This is not a valid acknowledgement decision, but fail closed
            // if the shared policy grows a new outcome without this adapter
            // being updated at the same time.
            return .authoritativeTerminal
        default:
            // Keep the adapter bounded if the KMP enum gains a new outcome:
            // never restore text or touch a newer action implicitly.
            return .authoritativeTerminal
        }
    }
}
